roles/lib_openshift/src/class/oc_configmap.py

```diff
@@ -0,0 +1,183 @@
+                 name,
+                 from_file,
+                 from_literal,
+                 state,
+                 namespace,
+                 kubeconfig='/etc/origin/master/admin.kubeconfig',
+                 verbose=False):
+        ''' Constructor for OpenshiftOC '''
+        super(OCConfigMap, self).__init__(namespace, kubeconfig=kubeconfig, verbose=verbose)
+        self.name = name
+        self.state = state
+        self._configmap = None
+        self._inc_configmap = None
+        self.from_file = from_file if from_file is not None else {}
+        self.from_literal = from_literal if from_literal is not None else {}
+
+    @property
+    def configmap(self):
+        if self._configmap is None:
+            self._configmap = self.get()
+
+        return self._configmap
+
+    @configmap.setter
+    def configmap(self, inc_map):
+        self._configmap = inc_map
+
+    @property
+    def inc_configmap(self):
+        if self._inc_configmap is None:
+            results = self.create(dryrun=True, output=True)
+            self._inc_configmap = results['results']
+
+        return self._inc_configmap
+
+    @inc_configmap.setter
+    def inc_configmap(self, inc_map):
+        self._inc_configmap = inc_map
+
+    def from_file_to_params(self):
+        '''return from_files in a string ready for cli'''
+        return ["--from-file={}={}".format(key, value) for key, value in self.from_file.items()]
+
+    def from_literal_to_params(self):
+        '''return from_literal in a string ready for cli'''
+        return ["--from-literal={}={}".format(key, value) for key, value in self.from_literal.items()]
+
+    def get(self):
+        '''return a configmap by name '''
+        results = self._get('configmap', self.name)
+        if results['returncode'] == 0 and results['results'][0]:
+            self.configmap = results['results'][0]
+
+        if results['returncode'] != 0 and '"{}" not found'.format(self.name) in results['stderr']:
+            results['returncode'] = 0
+
+        return results
+
+    def delete(self):
+        '''delete a configmap by name'''
+        return self._delete('configmap', self.name)
+
+    def create(self, dryrun=False, output=False):
+        '''Create a configmap
+
+           :dryrun: Product what you would have done. default: False
+           :output: Whether to parse output. default: False
+        '''
+
+        cmd = ['create', 'configmap', self.name]
+        if self.from_literal is not None:
+            cmd.extend(self.from_literal_to_params())
+
+        if self.from_file is not None:
+            cmd.extend(self.from_file_to_params())
+
+        if dryrun:
+            cmd.extend(['--dry-run', '-ojson'])
+
+        results = self.openshift_cmd(cmd, output=output)
+
+        return results
+
+    def update(self):
+        '''run update configmap '''
+        return self._replace_content('configmap', self.name, self.inc_configmap)
+
+    def needs_update(self):
```

Comment:
So to be clear, the previous statement about `patch` as it could be used today is not meant to be "we should always use `patch`". There are use-cases for `patch` and use-cases for `replace`. These are separate use-cases. I would argue that a `replace` in a re-try loop is a very suspect operation. What would the re-try loop look like?

 1. read object state from server
 2. make edit
 3. try to replace object state on server

When you pass resource version on the `replace` call and get a failure, it means someone else changed the object between steps 1 and 3. So there are a couple cases:

 1. you know the full and correct object state and mean with `oc replace` that the entirety of your YAML or JSON is correct and nothing else should be on the server
 2. you know some subset of correct object state and mean with `oc replace` to edit that portion of the object

For the first option, it does not matter what the other person edited between steps 1 and 3 ... you know the full and correct state. Pass no resource version and ensure that your object is on the server as you intend.

For the second option, the user making an edit between steps 1 and 3 could either change the fields you mean to edit, or unrelated fields. If they change unrelated fields, this means that in your step 2 you'll make a new version of the object with their edits and yours, then submit that to `oc replace`. This is the use-case where `oc patch` is what you really want to use, because the process of reading an object, patching it and re-introducing it into the server is what `oc patch` does... but you are able to encode your intent to change a field and the server can see that the other unrelated change won't impact you.

In the other case, the other unrelated user made a change to the field you meant to change and you either know that your field is always correct or sometimes correct. If you know that you have the correct values for the fields in your `patch` you should be fine to use `oc patch` whenever. If you only want to do the patch sometimes... then realistically you want `oc apply` with pre-conditions on your action. This is not yet implemented but you can expect it to be.